torchtitan/datasets/hf_datasets.py

```diff
@@ -107,36 +108,71 @@ def __init__(
 
         self.dataset_name = dataset_name
         self._data = split_dataset_by_node(ds, rank, world_size)
+        self._sample_idx = 0
         self._tokenizer = tokenizer
+        self._all_tokens: List[int] = []
         self.seq_len = seq_len
         self.infinite = infinite
+        self._first_iter = True
 
     def __iter__(self):
         max_buffer_token_len = 1 + self.seq_len
-        all_tokens: List[int] = []
 
         while True:
-            for sample in iter(self._data):
+            # Skip samples for the first iteration
+            data_iter = (
+                self._get_skipped_data_iter(self._sample_idx)
+                if self._first_iter
+                else iter(self._data)
+            )
+            self._first_iter = False
+
+            for sample in data_iter:
                 sample_text = sample["text"]
                 sample_tokens = self._tokenizer.encode(sample_text, bos=True, eos=True)
-                all_tokens.extend(sample_tokens)
+                self._all_tokens.extend(sample_tokens)
+                self._sample_idx += 1
 
-                while len(all_tokens) >= max_buffer_token_len:
-                    x = torch.LongTensor(all_tokens[:max_buffer_token_len])
+                while len(self._all_tokens) >= max_buffer_token_len:
+                    x = torch.LongTensor(self._all_tokens[:max_buffer_token_len])
                     # update tokens to the remaining tokens
-                    all_tokens = all_tokens[max_buffer_token_len:]
+                    self._all_tokens = self._all_tokens[max_buffer_token_len:]
                     input = x[:-1]
                     label = x[1:]
                     yield input, label
+
             if not self.infinite:
                 logger.warning(f"Dataset {self.dataset_name} has run out of data.")
                 break
             else:
+                # Reset offset for the next iteration
+                self._sample_idx = 0
                 logger.warning(
                     f"Dataset {self.dataset_name} is being re-looped. "
                     "Loss related metrics might be misleading."
                 )
 
+    def _get_skipped_data_iter(self, skip_samples):
+        if isinstance(self._data, IterableDataset):
+            it = iter(self._data)
+            # Naively iterate through the samples as dataset length is unknown
+            for _ in range(skip_samples):
+                next(it)
+            return it
+
+        # Map-style dataset
+        # As skipping to the end throws an error, return an empty iterator
+        if skip_samples == len(self._data):
+            return iter([])
+        return iter(self._data.skip(skip_samples))
+
+    def load_state_dict(self, state_dict):
+        self._sample_idx = state_dict["sample_idx"]
+        self._all_tokens = state_dict["token_buffer"]
+
+    def state_dict(self):
+        return {"token_buffer": self._all_tokens, "sample_idx": self._sample_idx}
```

Comment:
Is the `_all_tokens` a DTensor? If not, only rank0's dataloader data will be saved.

Reply:
Same as `_sample_idx`.

Reply:
@fegin For now, the state_dict returned by the dataloader is keyed by the rankid. That is {0: state_dict}. Would that ensure the state of all the ranks is stored? Whether we want to use DTensor is something we are looking into (as resharding dataloader state is a challenge anyway so the benefit of storing it as DTensor is a bit limited)

Reply:
As long as the key is unique across ranks, using DTensor is not required.